resource_estimation/Error_Budget_Optimization/generate_data.py

```diff
@@ -0,0 +1,139 @@
+
+        physicalqubits = result["physicalCounts"]["physicalQubits"]
+        runtime = result["physicalCounts"]["runtime"]
+
+        default_metric = default_runtime * default_physicalqubits
+        current_metric = runtime * physicalqubits
+
+        if running_metric is None:
+            running_metric = current_metric
+            logical_cache = logical_budget
+            t_cache = t_budget
+            rotation_cache = rotation_budget
+
+        if current_metric < running_metric:
+            running_metric = current_metric
+            logical_cache = logical_budget
+            t_cache = t_budget
+            rotation_cache = rotation_budget
+
+    running_optimal_parameters["logical_budget"] = logical_cache
+    running_optimal_parameters["t_budget"] = t_cache
+    running_optimal_parameters["rotation_budget"] = rotation_cache
+
+    return list(running_optimal_parameters.values()), running_metric, default_metric
+
+
+def generate_data(total_error_budget, counts, path="MQTBench"):
+    """
+    Generates a dataset consisting of logical counts of quantum circuits and respective optimized error budgets.
+
+    This function searches for QASM files in the "MQTBench" directory, loads each circuit,
+    estimates its logical counts, and computes optimized error budget partitions using random sampling.
+    For each circuit, it collects relevant counts and the optimal error budget distribution,
+    appending the results to a list.
+
+    Args:
+        total_error_budget: The total error budget to be distributed among error types.
+        counts: LogicalCounts object or dictionary containing logical counts for estimation.
+
+    Returns:
+        A list of lists, where each inner list contains circuit-specific counts and the corresponding
+        optimized error budget partition.
+    """
+    qasm_files = [Path(root) / file for root, _, files in os.walk(path) for file in files if file.endswith(".qasm")]
+    qasm_files = sorted(qasm_files)
+    print(qasm_files[-1])
+    results = []
+
+    for file in tqdm(qasm_files):
+        with Path.open(file, encoding="utf-8") as f:
+            qasm = f.read()
+            qc = QuantumCircuit.from_qasm_str(qasm)
+        try:
+            estimation = estimate(qc)
+            counts = estimation["logicalCounts"]
+            if counts["rotationCount"] == 0:
+                continue
+            counts = LogicalCounts(counts)
+            combinations, _running_metric, _default_metric = find_optimized_budgets(total_error_budget, 1000, counts)
+            specific_data = [
+                int(counts["numQubits"]),
+                int(counts["tCount"]),
+                int(counts["rotationCount"]),
+                int(counts["rotationDepth"]),
+                int(counts["cczCount"]),
+                int(counts["ccixCount"]),
+                int(counts["measurementCount"]),
+            ]
+            specific_data += combinations
+            results.append(specific_data)
```

Comment:
This is a fairly unstructured way to track the results. Something like a TypedDict or a data class might be more suitable here.